hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/common/model/HoodieSparkRecord.java

```diff
@@ -307,7 +313,14 @@ public Option<Map<String, String>> getMetadata() {
 
   @Override
   public Option<HoodieAvroIndexedRecord> toIndexedRecord(Schema recordSchema, Properties prop) {
-    throw new UnsupportedOperationException();
+    if (data == null) {
+      return Option.empty();
+    }
+    GenericRecord convertedRecord = AVRO_CONVERTER_CACHE.computeIfAbsent(Pair.of(schema, recordSchema), pair -> {
+      StructType structType = schema == null ? AvroConversionUtils.convertAvroSchemaToStructType(recordSchema) : schema;
+      return data -> AvroConversionUtils.createInternalRowToAvroConverter(structType, recordSchema, false).apply(data);
```

Comment:
can we also cache the nested 2 converters, like what we do in `HoodieInternalRowUtils`

Reply:
Which converters are you referring to?

Reply:
these two:
1. `AvroConversionUtils.createInternalRowToAvroConverter(structType, recordSchema, false)`
2. `AvroConversionUtils.convertAvroSchemaToStructType(recordSchema)`

Reply:
PR is updated